Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) Research In Motion Limited 2009-2010. All rights reserved.
+ * Copyright (C) 2021, 2022, 2023 Igalia S.L.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "RenderSVGResourceMasker.h"
+
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+#include "Element.h"
+#include "ElementIterator.h"
+#include "FloatPoint.h"
+#include "Image.h"
+#include "ImageBuffer.h"
+#include "IntRect.h"
+#include "RenderLayer.h"
+#include "RenderLayerInlines.h"
+#include "RenderSVGResourceMaskerInlines.h"
+#include "SVGContainerLayout.h"
+#include "SVGLengthContext.h"
+#include "SVGRenderStyle.h"
+#include "SVGRenderingContext.h"
+#include "SVGResources.h"
+#include "SVGResourcesCache.h"
+#include <wtf/IsoMallocInlines.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(RenderSVGResourceMasker);
+
+RenderSVGResourceMasker::RenderSVGResourceMasker(SVGMaskElement& element, RenderStyle&& style)
+    : RenderSVGResourceContainer(Type::SVGResourceMasker, element, WTFMove(style))
+{
+}
+
+RenderSVGResourceMasker::~RenderSVGResourceMasker() = default;
+
+static RefPtr<ImageBuffer> createImageBuffer(const FloatRect& targetRect, const AffineTransform& absoluteTransform, const DestinationColorSpace& colorSpace, const GraphicsContext* context)
+{
+    IntRect paintRect = enclosingIntRect(absoluteTransform.mapRect(targetRect));
+    // Don't create empty ImageBuffers.
+    if (paintRect.isEmpty())
+        return nullptr;
+
+    FloatSize scale;
+    FloatSize clampedSize = ImageBuffer::clampedSize(paintRect.size(), scale);
+
+#if USE(DIRECT2D)
+    auto imageBuffer = ImageBuffer::create(clampedSize, RenderingPurpose::Unspecified, context, 1, colorSpace, PixelFormat::BGRA8);
+#else
+    UNUSED_PARAM(context);
+    auto imageBuffer = ImageBuffer::create(clampedSize, RenderingPurpose::Unspecified, 1, colorSpace, PixelFormat::BGRA8);
+#endif
+    if (!imageBuffer)
+        return nullptr;
+
+    AffineTransform transform;
+    transform.scale(scale).translate(-paintRect.location()).multiply(absoluteTransform);
```

Comment:
IIRC we have a generic function that can be used to create image buffers. Maybe I'm wrong. Let's check :-)

Reply:
I tried quickly and saw a regression in svg/custom (did not try other directories). Since this could be some subtle code difference we agreed to leave this for later.